Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/storage/repo/DefaultHostedEventRepository.kt

```diff
@@ -0,0 +1,54 @@
+package de.rki.coronawarnapp.eventregistration.storage.repo
+
+import de.rki.coronawarnapp.eventregistration.events.HostedEvent
+import de.rki.coronawarnapp.eventregistration.storage.EventRegistrationDatabase
+import de.rki.coronawarnapp.eventregistration.storage.dao.HostedEventDao
+import de.rki.coronawarnapp.eventregistration.storage.entity.toHostedEventEntity
+import de.rki.coronawarnapp.util.coroutine.AppScope
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class DefaultHostedEventRepository @Inject constructor(
+    eventRegistrationDatabaseFactory: EventRegistrationDatabase.Factory,
+    @AppScope private val appScope: CoroutineScope
+) : HostedEventRepository {
+
+    private val eventRegistrationDatabase: EventRegistrationDatabase by lazy {
+        eventRegistrationDatabaseFactory.create()
+    }
+
+    private val hostedEventDao: HostedEventDao by lazy {
+        eventRegistrationDatabase.hostedEventsDao()
+    }
+
+    override val allHostedEvents: Flow<List<HostedEvent>>
+        get() = hostedEventDao.allEntries() // TODO: SORTING
+
+    override fun addHostedEvent(event: HostedEvent) {
```

Comment:
is not Room already support `suspend`?

Reply:
Yes, the dao function is a suspend function. I am using the appscope here so that the coroutine is not canceled if the calling scope (viewModelScope) is canceled when the user leaves the screen. So we make sure that the data is really written into the Database and not canceled prematurely.

Reply:
Does this mean that all DAOs we have are not working well? What I know that room uses its own dispatcher

Reply:
Ah, this is in the repo not the DAO